src/lib/support/CodeUtils.h

```diff
@@ -388,9 +389,12 @@ constexpr inline const _T & max(const _T & a, const _T & b)
  *  @endcode
  *
  *  @param[in]  aStatus     A scalar status to be evaluated against zero (0).
+ *  @param[in]  aMetricKey  An optional metric key to be logged if the condition
+ *                          evaluates to false. The value for the metric is
+ *                          result of the expression aStatus.
  *
  */
-#define SuccessOrExit(aStatus) nlEXPECT(::chip::ChipError::IsSuccess((aStatus)), exit)
+#define SuccessOrExit(aStatus, ...) nlEXPECT(LOG_METRIC_FOR_SUCCESS_OR_EXIT((aStatus), ##__VA_ARGS__), exit)
```

Comment:
We also have `ReturnErrorOnFailure` for example ... should those also be affected?

Overall I find vararg macros with optional parameters very hard to reason about. Separate naming or explicit wrappers (some `CHIPError &LoggedIfError(CHIPError &e)` for example) that we place when needed seems better.

If we modify macros, a single error seems to also metric out through the entire stack, which may be too much.

Reply:
I feel quite strongly FYI that using existing paradigms is key to our success here and having an optional second parameter is not rocket science here, it fits quite nicely, and teaches people about metrics.

Reply:
Please see https://github.com/project-chip/connectedhomeip/pull/32223/files#r1497141286

Reply:
> I feel quite strongly FYI that using existing paradigms is key to our success here and having an optional second parameter is not rocket science here, it fits quite nicely, and teaches people about metrics.

I don't dispute that there is a strong feeling. I am saying that the naming is key, and it would be easy to make a new `TracingSuccessOrExit`, and this would avoid having macros where we can "miss" the fact that there is a metric key at the end. It's really not obvious when you scan code, given that the metric is "late in the line" and man y of the `SuccessOrExit` lines are very very long (sometimes multiline, as in this PR).

See https://github.com/project-chip/connectedhomeip/pull/32223/files#r1497693518.

Having new macros that exist in tracing.h, also would split the dep, in a way that

CodeUtils.h doesn't depend on tracing.h, but rather Tracing.h depends on CodeUtils. That means if you start having SuccessOrXxxx in Tracing.h for other reasons, for example, in inline methods, you will not have circular deps.

Reply:
I think this is resolved